config.org
#+TITLE: Emacs Configuration
#+AUTHOR: Shantanu Bhardwaj

I have been using Emacs since winter of 2000 at the The Ohio State University
and i have declared Emacs Bankruptcy many times. It  has always been a learning experience.
Mostly it works best when I stick with it.

* Personal Information

#+begin_src emacs-lisp
  (setq user-full-name "Shantanu Bhardwaj"
        user-mail-address "shan@addvalsolutions.com")
#+end_src

* Setup Packages

#+begin_src emacs-lisp
  ;; Add only the MELPA archive
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("elpa" . "https://elpa.gnu.org/packages/")))

  (package-initialize)

  (require 'use-package)

  (use-package exec-path-from-shell
    :ensure t
    :config
    (exec-path-from-shell-initialize))

  ;; Configure a package
  (use-package org
    :ensure t
    :config
    (setq org-log-done 'time))

  (use-package auto-package-update
    :defer t
    :custom
    (setq auto-package-update-interval 7
          auto-package-update-prompt-before-update t
          auto-package-update-hide-results nil))

  (let* ((dir (expand-file-name (concat user-emacs-directory "local-packages")))
         (default-directory dir))
    (when (file-directory-p dir)
      (add-to-list 'load-path dir)
      (if (fboundp 'normal-top-level-add-subdirs-to-load-path)
          (normal-top-level-add-subdirs-to-load-path))))

#+end_src
* Customize settings

Set up the customize file to its own separate file, instead of saving
customize settings in [[file:init.el][init.el]].

#+begin_src emacs-lisp
  (setq max-lisp-eval-depth 10000)
  (setq custom-file (expand-file-name "custom.el" user-emacs-directory))
  (load custom-file)
  (menu-bar-mode -1)
  (display-battery-mode -1)        ;; Show battery.
  (display-time-mode t)           ;; Show time.
  (fset 'yes-or-no-p 'y-or-n-p)   ;; Set yes or no to y/n
  (global-auto-revert-mode)       ;; Refresh a buffer if changed on disk
  (global-hl-line-mode 1)         ;; Highlight current line
  (blink-cursor-mode -1)          ;; Disable blingking cursor
  (recentf-mode t)                ;; Turn on recent files mode

  ;; (pixel-scroll-precision-mode 1)
  ;; Enable pixel-based scrolling
  ;; (require 'pixel-scroll)
  ;; (pixel-scroll-mode 1)

  ;; Customize pixel-based scrolling
  ;; (setq pixel-scroll-precision-large-scroll-height 10)
  ;; (setq pixel-scroll-precision-interpolate-page t)
  ;; (setq pixel-scroll-precision-use-momentum t)

  (setq ad-redefinition-action            'accept
        global-auto-revert-non-file-buffers t
        auto-revert-check-vc-info         t
        backup-by-copying                 t
        backup-directory-alist            '(("." . "~/.emacs.d/backups"))
        cursor-in-non-selected-windows    nil
        byte-compile-warnings             '(ck-functions)
        confirm-kill-processes            nil
        create-lockfiles                  nil
        echo-keystrokes                   0.2
        column-number-mode                t
        find-file-visit-truename          t
        font-lock-maximum-decoration      t
        highlight-nonselected-windows     t
        fast-but-imprecise-scrolling      t
        jit-lock-defer-time               nil
        kill-buffer-query-functions       nil    ;; Dont ask for closing spawned processes
        scroll-margin                     0   ;; scroll N to screen edge
        load-prefer-newer                 t
        use-dialog-box                    nil
        visible-bell                      nil
        word-wrap                         nil
        auto-mode-case-fold               nil
        truncate-lines                    t
        truncate-string-ellipsis          ".."
        undo-limit                        6710886400 ;; 64mb
        undo-strong-limit                 100663296 ;; x 1.5 (96mb)
        undo-outer-limit                  1006632960) ;; x 10 (960mb), (Emacs uses x100), but this seems too high.

#+end_src

* Custom functions
#+begin_src emacs-lisp
  (defun align-values (start end)
    "Vertically aligns region based on lengths of the first value of each line.
  Example output:

          foo        bar
          foofoo     bar
          foofoofoo  bar"
    (interactive "r")
    (align-regexp start end
                  "\\([a-z_]+: \\)"
                  -1 1 nil))
#+end_src

#+begin_src emacs-lisp
  ;; (defun fontify-frame (frame)
  ;;   (interactive)
  ;;   (if sys/macp
  ;;       (progn
  ;;         (if (> (x-display-pixel-width) 4000)
  ;;             (set-frame-parameter frame 'font "Monaco 14") ;; 5k Display
  ;;           (set-frame-parameter frame 'font "Monaco 12")))))

  ;; ;; Fontify current frame
  ;; (fontify-frame nil)

  ;; ;; Fontify any future frames
  ;; (push 'fontify-frame after-make-frame-functions)

#+end_src
* UI + Theme
** Icons
#+begin_src emacs-lisp
(use-package spacious-padding
  :ensure t
  :config
  ;; (setq spacious-padding-widths
  ;;       '( :internal-border-width 15
  ;;          :header-line-width 4
  ;;          :mode-line-width 6
  ;;          :tab-width 4
  ;;          :right-divider-width 30
  ;;          :scroll-bar-width 8))
  (spacious-padding-mode 1)
  )

;; Install and configure modus-themes
(use-package ef-themes
  :ensure t
  :config
  (load-theme 'ef-spring t)
  )

(use-package auto-dark
  :ensure t
  :custom
  (auto-dark-themes '((ef-dream) (ef-spring)))
  (auto-dark-polling-interval-seconds 300)
  (auto-dark-allow-osascript t)
  (auto-dark-allow-powershell nil)
  ;; (auto-dark-detection-method nil) ;; dangerous to be set manually
  :hook
  (auto-dark-dark-mode
   . (lambda ()
       ;; something to execute when dark mode is detected
       ))
  (auto-dark-light-mode
   . (lambda ()
       ;; something to execute when light mode is detected
       ))
  :init (auto-dark-mode))

(use-package nerd-icons
  :ensure t)

(use-package nerd-icons-completion
  :ensure t
  :after marginalia
  ;; FIXME 2024-09-01: For some reason this stopped working because it
  ;; macroexpands to `marginalia-mode' instead of
  ;; `marginalia-mode-hook'.  What is more puzzling is that this does
  ;; not happen in the next :hook...
  ;; :hook (marginalia-mode . nerd-icons-completion-marginalia-setup))
  :config
  (add-hook 'marginalia-mode-hook #'nerd-icons-completion-marginalia-setup))

(use-package nerd-icons-dired
  :hook (dired-mode . nerd-icons-dired-mode))

(use-package nerd-icons-corfu
  :ensure t
  :after corfu
  :config
  (add-to-list 'corfu-margin-formatters #'nerd-icons-corfu-formatter)
  ;; Optionally:
  (setq nerd-icons-corfu-mapping
        '((array :style "cod" :icon "symbol_array" :face font-lock-type-face)
          (boolean :style "cod" :icon "symbol_boolean" :face font-lock-builtin-face)
          ;; ...
          (t :style "cod" :icon "code" :face font-lock-warning-face)))
  ;; Remember to add an entry for `t', the library uses that as default.
  )

#+end_src
** Convenient theme functions

#+begin_src emacs-lisp

  ;;  (load-theme 'gruvbox-dark-hard t)

  (defun reset-modeline()
    "Reset the modeline to 12pt font"
    (interactive)
    (set-face-attribute 'mode-line nil :height 100)
    (set-face-attribute 'mode-line-inactive nil :height 100)
    (setq doom-modeline-height 12)
    (set-face-attribute 'line-number nil :font "Monaco-10")
    (set-face-attribute 'line-number-current-line nil :font "Monaco-10")
    )

  (defun switch-theme (theme)
    "Disables any currently active themes and loads THEME."
    ;; This interactive call is taken from `load-theme'
    (interactive
     (list
      (intern (completing-read "Load custom theme: "
                               (mapc 'symbol-name
                                     (custom-available-themes))))))
    (let ((enabled-themes custom-enabled-themes))
      (mapc #'disable-theme custom-enabled-themes)
      (load-theme theme t)
      (reset-modeline)
      ))

(defun disable-active-themes ()
  "Disables any currently active themes listed in `custom-enabled-themes'."
  (interactive)
  (mapc #'disable-theme custom-enabled-themes))

(bind-key "s-<f12>" 'consult-theme)
(bind-key "s-<f11>" 'disable-active-themes)
#+end_src
** Modeline

#+begin_src emacs-lisp
  (use-package doom-modeline
    :hook (after-init . doom-modeline-mode)
    :custom    
    (doom-modeline-height 20)
    (doom-modeline-bar-width 1)
    (doom-modeline-icon t)
    (doom-modeline-major-mode-icon t)
    (doom-modeline-major-mode-color-icon t)
    (doom-modeline-time-analogue-clock nil)
    (doom-modeline-buffer-file-name-style 'truncate-upto-project)
    (doom-modeline-buffer-state-icon t)
    (doom-modeline-buffer-modification-icon t)
    (doom-modeline-minor-modes nil)
    (doom-modeline-enable-word-count nil)
    (doom-modeline-buffer-encoding t)
    (doom-modeline-indent-info nil)
    (doom-modeline-checker-simple-format t)
    (doom-modeline-vcs-max-length 12)
    (doom-modeline-env-version t)
    (doom-modeline-irc-stylize 'identity)
    (setq doom-modeline-github t)
    (setq doom-modeline-battery-icon nil)
    (setq doom-modeline-mu4e t)
    (doom-modeline-github-timer nil)
    (doom-modeline-gnus-timer nil))

#+end_src
** Font

#+begin_src emacs-lisp
;; (set-face-attribute 'default nil :font "Iosevka Comfy" :height 140 :weight 'regular) 
(set-face-attribute 'default nil :font "CaskaydiaMono Nerd Font" :height 140 :weight 'regular) 


#+end_src

Note: To view all fonts that are available to use, run the following:

#+BEGIN_SRC emacs-lisp :tangle no
  (font-family-list)
#+END_SRC

** Custom Mode faces

*** Echo Area
#+begin_src emacs-lisp
  (with-current-buffer " *Echo Area 0*" (face-remap-add-relative 'default '(:family "Monaco" :height 110)))
#+end_src

*** Term Faces

#+begin_src emacs-lisp
  ;; Use monospaced font faces in current buffer
  (defun my-term-mode-face ()
    "Sets a fixed width (monospace) font in current buffer"
    (interactive)
    (setq buffer-face-mode-face '(:family "Monaco" :height 100))
    (buffer-face-mode))

#+end_src


* Sane defaults

#+begin_src emacs-lisp
  (setq-default display-line-numbers-width    3       ;; Set so we can display thousands of lines
                c-basic-offset                2            ;; Set tab indent for c/c++ to 4 tabs
                ediff-forward-word-function   'forward-char
                ediff-split-window-function   'split-window-horizontally
                tab-width                     2            ;: Use four tabs
                indent-tabs-mode              nil			 ;; Never use tabs. Use spaces instead
                truncate-lines                t
                indent-line-function          'insert-tab  ;; Use function to insert tabs
                history-length                100
                uniquify-buffer-name-style    'forward)

  ;; These functions are useful. Activate them.
  (put 'downcase-region 'disabled nil)
  (put 'upcase-region 'disabled nil)
  (put 'narrow-to-region 'disabled nil)
  (put 'dired-find-alternate-file 'disabled nil)

  ;; UTF-8 please
  (setq locale-coding-system 'utf-8) ; pretty
  (set-terminal-coding-system 'utf-8) ; pretty
  (set-keyboard-coding-system 'utf-8) ; pretty
  (set-selection-coding-system 'utf-8) ; please
  (prefer-coding-system 'utf-8) ; with sugar on top
  (setq-default indent-tabs-mode nil)

  (setq-default indent-tabs-mode nil)
  (setq-default indicate-empty-lines t)

  ;; Don't count two spaces after a period as the end of a sentence.
  ;; Just one space is needed.
  (setq sentence-end-double-space nil)

  ;; delete the region when typing, just like as we expect nowadays.
  (delete-selection-mode t)
  (show-paren-mode t)
  (column-number-mode t)

  ;; (global-visual-line-mode -1)
  (remove-hook 'text-mode-hook #'turn-on-auto-fill)
  (add-hook 'text-mode-hook 'turn-on-visual-line-mode)
  (diminish 'visual-line-mode)

  (setq uniquify-buffer-name-style 'forward)

  ;; -i gets alias definitions from .bash_profile
  (setq shell-command-switch "-ic")

  (global-unset-key (kbd "M-m"))
  (global-set-key (kbd "C-+") 'text-scale-increase)
  (global-set-key (kbd "C--") 'text-scale-decrease)
  ;; Bind C-c i to open init.el
  (global-set-key (kbd "C-c e") (lambda () (interactive) (find-file "~/.emacs.d/init.el")))


  (windmove-default-keybindings 'super)
  (setq native-comp-async-report-warnings-errors 'silent)
#+end_src


* Mac customizations

There are configurations to make when running Emacs on macOS (hence the
"darwin" system-type check).

#+begin_src emacs-lisp
  (when (string-equal system-type "darwin")

  (setq mac-option-key-is-meta nil
        mac-command-key-is-meta t
        mac-command-modifier 'meta
        mac-option-modifier 'super
        mac-function-modifier 'hyper
        dired-use-ls-dired nil)

  ;; Fullscreen!
  (setq ns-use-native-fullscreen nil) ; Not Lion style
  (bind-key "<s-return>" 'toggle-frame-fullscreen)

  ;; buffer switching
  (bind-key "s-[" 'previous-buffer)
  (bind-key "s-]" 'next-buffer)

  (bind-key "M-C-w" 'restart-emacs)


    ;; delete files by moving them to the trash
    (setq delete-by-moving-to-trash t)
    (setq trash-directory "~/.Trash")
    (setq ns-pop-up-frames nil)

    ;; Compiling
    (bind-key "H-c" 'compile)
    (bind-key "H-r" 'recompile)
    (bind-key "H-s" (defun save-and-recompile () (interactive) (save-buffer) (recompile))))
#+end_src


* MELPA / ELPA

** DAbbrev
#+begin_src emacs-lisp

(use-package dabbrev
  :ensure nil
  :custom
  (dabbrev-ignored-buffer-regexps '("\\.\\(?:pdf\\|jpe?g\\|png\\)\\'")))


#+end_src
** Expand Region
#+begin_src emacs-lisp
(use-package expand-region
  :bind ("C-=" . er/expand-region))

#+end_src

** Indent
#+begin_src emacs-lisp

(use-package indent-bars
  :ensure nil
  :config
  (setq
   indent-bars-color '(highlight :face-bg t :blend 0.3)
   indent-bars-pattern " . . . . ." ; play with the number of dots for your usual font size
   indent-bars-width-frac 0.25
   indent-bars-pad-frac 0.1)
  :custom
  (indent-bars-treesit-support t)
  (indent-bars-no-descend-string t)
  (indent-bars-treesit-ignore-blank-lines-types '("module"))
  (indent-bars-treesit-wrap '((python argument_list parameters ; for python, as an example
				                              list list_comprehension
				                              dictionary dictionary_comprehension
				                              parenthesized_expression subscript)))
  :hook (prog-mode . indent-bars-mode))

#+end_src

** SVG Tags
#+begin_src emacs-lisp
;; (use-package svg-tag-mode
;;   :hook (prog-mode . svg-tag-mode)
;;   :config
;;   (plist-put svg-lib-style-default :font-family "Monaco Nerd Font Mono")
;;   (plist-put svg-lib-style-default :font-size 14)
;;   ;; (require 'periphery)
;;   ;; (setq svg-tag-tags (periphery-svg-tags))
;;   )

#+end_src

** Misc
#+begin_src emacs-lisp
(use-package toml)
(use-package gcmh
  :diminish
  :hook (after-init . gcmh-mode))
(use-package diminish
  :ensure t)


#+end_src

** Helpful
#+begin_src emacs-lisp
(use-package helpful
  :ensure t
  :commands (helpful-callable helpful-variable helpful-command helpful-key)
  :bind (("C-h f" . helpful-callable)
         ("C-h v" . helpful-variable)
         ("C-h k" . helpful-key)
         ("C-h x" . helpful-command)
         ("C-c C-d" . helpful-at-point)
         ("C-c F" . helpful-function)))

#+end_src

** HL Todo
#+begin_src emacs-lisp
(use-package hl-todo
  :ensure t
  :defer t
  :custom-face (hl-todo ((t (:box t :inherit))))
  :bind (:map hl-todo-mode-map
              ([C-f3] . hl-todo-occur)
              ("C-c t p" . hl-todo-previous)
              ("C-c t n" . hl-todo-next)
              ("C-c t o" . hl-todo-occur))
  :hook (after-init . global-hl-todo-mode))

#+end_src

** No Littering
#+begin_src emacs-lisp
(use-package no-littering
  :ensure t
  :config
  (setq custom-file (expand-file-name "custom.el" user-emacs-directory))
  (require 'recentf)
  (add-to-list 'recentf-exclude
               (recentf-expand-file-name no-littering-var-directory))
  (add-to-list 'recentf-exclude
               (recentf-expand-file-name no-littering-etc-directory))
  ) 

#+end_src

** Multiple cursors

We'll also need to ~(require 'multiple-cusors)~ because of [[https://github.com/magnars/multiple-cursors.el/issues/105][an autoload issue]].

#+begin_src emacs-lisp
  (use-package multiple-cursors
    :ensure t
    :defer t
    :bind (("C-S-c C-S-c" . mc/edit-lines)
           ("C->"         . mc/mark-next-like-this)
           ("C-<"         . mc/mark-previous-like-this)
           ("C-c C-<"     . mc/mark-all-like-this)
           ("C-!"         . mc/mark-next-symbol-like-this)
           ("s-d"         . mc/mark-all-dwim)))
#+end_src

** Rainbow mode + delimiters
#+BEGIN_SRC emacs-lisp
  (use-package rainbow-delimiters
    :ensure t
    :diminish
    :hook (prog-mode . rainbow-delimiters-mode))

  (use-package rainbow-mode
    :ensure t
    :diminish
    :hook (prog-mode . rainbow-mode)
    :config
    (setq rainbow-x-colors nil))

#+END_SRC

** Savehist
#+begin_src emacs-lisp
;; Further reading: https://protesilaos.com/emacs/dotemacs#h:25765797-27a5-431e-8aa4-cc890a6a913a
;;;; `savehist' (minibuffer and related histories)
(use-package savehist
  :ensure nil
  :hook (after-init . savehist-mode)
  :config
  (setq savehist-file (locate-user-emacs-file "savehist"))
  (setq history-length 100)
  (setq history-delete-duplicates t)
  (setq savehist-save-minibuffer-history t)
  (add-to-list 'savehist-additional-variables 'kill-ring))

#+end_src

** Smoothscrolling

This makes it so ~C-n~-ing and ~C-p~-ing won't make the buffer jump
around so much.

#+begin_src emacs-lisp
  (use-package smooth-scrolling
    :ensure t)
#+end_src

** Scratch

Convenient package to create =*scratch*= buffers that are based on the
current buffer's major mode. This is more convienent than manually
creating a buffer to do some scratch work or reusing the initial
=*scratch*= buffer.

#+begin_src emacs-lisp
  (use-package scratch
    :ensure t
    :commands scratch)

  (use-package persistent-scratch
    :ensure t
    :init
    (persistent-scratch-setup-default))
#+end_src

** Smartparens
#+BEGIN_SRC emacs-lisp
  (use-package smartparens
    :ensure t
    :defer t
    :diminish smartparens-mode
    :hook (prog-mode . smartparens-mode))
#+END_SRC

** Super Save
#+BEGIN_SRC emacs-lisp
  (use-package super-save
    :ensure t
    :defer t
    :config
    ;; (super-save-mode +1)
    )
#+END_SRC
** Undo Redo

#+BEGIN_SRC emacs-lisp
(use-package undo-fu
  :ensure t
  :config
  (setq undo-fu-allow-undo-in-region t)
  (global-unset-key (kbd "M-z"))
  (global-set-key (kbd "M-z")   'undo-fu-only-undo)
  (global-set-key (kbd "M-S-z") 'undo-fu-only-redo))

(global-unset-key (kbd "C-z"))

(use-package vundo
  :ensure t
  :bind ("C-M-z" . vundo))

;; (set-face-attribute 'mode-line nil
;;                     :background "LightSteelBlue1"
;;                     :foreground "black"
;;                     :box "SkyBlue2")

  ;; (use-package undo-tree
  ;;   :ensure t
  ;;   :defer t
  ;;   :diminish
  ;;   :init
  ;;   (global-undo-tree-mode))
#+END_SRC
** Which Key
# TODO: Needs configuration
#+BEGIN_SRC emacs-lisp
  (use-package which-key
    :ensure t
    :defer t
    :diminish which-key-mode
    :hook (after-init . which-key-mode))

#+END_SRC
** Wgrep
#+begin_src emacs-lisp
;; The `wgrep' packages lets us edit the results of a grep search
;; while inside a `grep-mode' buffer.  All we need is to toggle the
;; editable mode, make the changes, and then type C-c C-c to confirm
;; or C-c C-k to abort.
;;
;; Further reading: https://protesilaos.com/emacs/dotemacs#h:9a3581df-ab18-4266-815e-2edd7f7e4852
(use-package wgrep
  :ensure t
  :bind ( :map grep-mode-map
          ("e" . wgrep-change-to-wgrep-mode)
          ("C-x C-q" . wgrep-change-to-wgrep-mode)
          ("C-c C-c" . wgrep-finish-edit)))

#+end_src

* Org mode
** Installation
#+BEGIN_SRC emacs-lisp
(use-package org
  :config
  (require 'org-tempo)
  )

(use-package org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
    :config
    (setq org-auto-tangle-babel-safelist '(
                                           "~/.emacs.d/config.org"
                                           )))
#+END_SRC
** Org activation bindings

Set up some global key bindings that integrate with Org Mode features.

#+begin_src emacs-lisp
  (bind-key "C-c l" 'org-store-link)
  (bind-key "C-c c" 'org-capture)
  (bind-key "C-c a" 'org-agenda)
#+end_src

*** Org agenda

Learned about [[https://github.com/sachac/.emacs.d/blob/83d21e473368adb1f63e582a6595450fcd0e787c/Sacha.org#org-agenda][this =delq= and =mapcar= trick from Sacha Chua's config]].

#+begin_src emacs-lisp
  (setq org-agenda-files
        (delq nil
              (mapcar (lambda (x) (and (file-exists-p x) x))
                      '("~/Documents/Agenda"))))
#+end_src

*** Org capture

#+begin_src emacs-lisp
  (bind-key "C-c c" 'org-capture)
  (setq org-default-notes-file "~/Documents/Notes/notes.org")
#+end_src

** Org setup

Speed commands are a nice and quick way to perform certain actions
while at the beginning of a heading. It's not activated by default.


eshell/history
exit
exit
exit
ls
mv new.html.erb rooms
mv edit.html.erb rooms
ls
exi 
exit 
exit 
extit
exit
EXIT
exit
cd ../../rumbl/
mix phx.gen.context Accounts Credential credentials email:string:unique password_hash:string user_id:references:users
cd ../../
mix phx.gen.context Accounts Credential credentials email:string:unique password_hash:string user_id:references:users
exit
cd ../../../
mix ecto.migrate 
exit
exit 
exit
exit
exit
exit
exit
exit
exit
cd projects/elixir/
ls
git clone git@github.com:korolvs/el_chat.git
exit
exit
exit 
exit
exit
exit
exit
createdb shan
psql -U shan -l
exit
cd ~/Downloads/Ruby_test
ls
ruby ruby_script.rb 
ls 
ruby ruby_script.rb 
ruby ruby_script.rb 
ruby ruby_script.rb 
ruby ruby_script.rb 
ruby ruby_script.rb 
ollama list 
exit


config.el
(setq user-full-name "Shantanu Bhardwaj"
      user-mail-address "shan@addvalsolutions.com")

(setq max-lisp-eval-depth 10000)
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(load custom-file)
(menu-bar-mode -1)

(defun align-values (start end)
  "Vertically aligns region based on lengths of the first value of each line.
Example output:

        foo        bar
        foofoo     bar
        foofoofoo  bar"
  (interactive "r")
  (align-regexp start end
                "\\([a-z_]+: \\)"
                -1 1 nil))

;; (defun fontify-frame (frame)
;;   (interactive)
;;   (if sys/macp
;;       (progn
;;         (if (> (x-display-pixel-width) 4000)
;;             (set-frame-parameter frame 'font "Monaco 14") ;; 5k Display
;;           (set-frame-parameter frame 'font "Monaco 12")))))

;; ;; Fontify current frame
;; (fontify-frame nil)

;; ;; Fontify any future frames
;; (push 'fontify-frame after-make-frame-functions)

;;  (load-theme 'gruvbox-dark-hard t)
  
  (defun reset-modeline()
    "Reset the modeline to 12pt font"
    (interactive)
    (set-face-attribute 'mode-line nil :height 100)
    (set-face-attribute 'mode-line-inactive nil :height 100)
    (setq doom-modeline-height 12)
    (set-face-attribute 'line-number nil :font "Monaco-10")
    (set-face-attribute 'line-number-current-line nil :font "Monaco-10")
    )

  (defun switch-theme (theme)
    "Disables any currently active themes and loads THEME."
    ;; This interactive call is taken from `load-theme'
    (interactive
     (list
      (intern (completing-read "Load custom theme: "
                               (mapc 'symbol-name
                                     (custom-available-themes))))))
    (let ((enabled-themes custom-enabled-themes))
      (mapc #'disable-theme custom-enabled-themes)
      (load-theme theme t)
      (reset-modeline)
      ))

  (defun disable-active-themes ()
    "Disables any currently active themes listed in `custom-enabled-themes'."
    (interactive)
    (mapc #'disable-theme custom-enabled-themes))

  (bind-key "s-<f12>" 'switch-theme)
  (bind-key "s-<f11>" 'disable-active-themes)

(add-to-list 'default-frame-alist
             ;; '(font . "-*-Operator Mono-medium-normal-normal-*-14-*-*-*-m-0-iso10646-1")
             '(font . "Monaco Nerd Font Mono-16")
             ;; '(font . "Cascadia Mono-14")
             )

(with-current-buffer " *Echo Area 0*" (face-remap-add-relative 'default '(:family "Monaco" :height 110)))

;; Use monospaced font faces in current buffer
(defun my-term-mode-face ()
  "Sets a fixed width (monospace) font in current buffer"
  (interactive)
  (setq buffer-face-mode-face '(:family "Monaco" :height 100))
  (buffer-face-mode))

;; These functions are useful. Activate them.
 (put 'downcase-region 'disabled nil)
 (put 'upcase-region 'disabled nil)
 (put 'narrow-to-region 'disabled nil)
 (put 'dired-find-alternate-file 'disabled nil)

 ;; Answering just 'y' or 'n' will do
 (defalias 'yes-or-no-p 'y-or-n-p)

 ;; Keep all backup and auto-save files in one directory
 ;; (setq backup-directory-alist '(("*" . "~/.emacs.d/backups")))
 ;; (setq auto-save-file-name-transforms '(("." "~/.emacs.d/auto-save-list/" t)))
 (setq make-backup-files nil)

 ;; UTF-8 please
 (setq locale-coding-system 'utf-8) ; pretty
 (set-terminal-coding-system 'utf-8) ; pretty
 (set-keyboard-coding-system 'utf-8) ; pretty
 (set-selection-coding-system 'utf-8) ; please
 (prefer-coding-system 'utf-8) ; with sugar on top
 (setq-default indent-tabs-mode nil)

 ;; Turn off the blinking cursor
 ;; (blink-cursor-mode -1)

 (setq-default indent-tabs-mode nil)
 (setq-default indicate-empty-lines t)

 ;; Don't count two spaces after a period as the end of a sentence.
 ;; Just one space is needed.
 (setq sentence-end-double-space nil)

 ;; delete the region when typing, just like as we expect nowadays.
 (delete-selection-mode t)

 (show-paren-mode t)

 (column-number-mode t)

 ;; (global-visual-line-mode -1)
 (remove-hook 'text-mode-hook #'turn-on-auto-fill)
 (add-hook 'text-mode-hook 'turn-on-visual-line-mode)
 (diminish 'visual-line-mode)

 (setq uniquify-buffer-name-style 'forward)

 ;; -i gets alias definitions from .bash_profile
 (setq shell-command-switch "-ic")

 ;; Don't beep at me
 (setq visible-bell nil)

 ;; highlight current line everywhere
 (global-hl-line-mode 1)

 (global-unset-key (kbd "M-m"))
 (global-set-key (kbd "C-+") 'text-scale-increase)
 (global-set-key (kbd "C--") 'text-scale-decrease)
                   ;; (global-set-key (kbd "C-+") 'text-scale-increase)
                   ;; (global-set-key (kbd "C--") 'text-scale-decrease)

 (add-hook 'prog-mode-hook #'display-line-numbers-mode)
 (windmove-default-keybindings 'super)
(setq native-comp-async-report-warnings-errors 'silent)

(when (string-equal system-type "darwin")

  (add-to-list 'default-frame-alist
               '(ns-transparent-titlebar . t))
  ;; set the window frame to dark theme
  (add-to-list 'default-frame-alist
               '(ns-appearance . dark))

  ;; delete files by moving them to the trash
  (setq delete-by-moving-to-trash t)
  (setq trash-directory "~/.Trash")
  ;; Don't make new frames when opening a new file with Emacs
  (setq ns-pop-up-frames nil)

  ;; set the Fn key as the hyper key
  ;; (setq ns-function-modifier 'hyper)
  ;; (setq ns-option-modifier 'super)
  ;; (setq ns-command-modifier 'meta)

  (setq mac-option-modifier 'super)
  (setq mac-command-modifier 'meta)

  ;; typical mac bindings
  (global-set-key (kbd "M-s") 'save-buffer)
  (global-set-key (kbd "M-z") 'undo)

  ;; Use Command-` to switch between Emacs windows (not frames)
  (bind-key "A-`" 'other-window)

  ;; Use Command-Shift-` to switch Emacs frames in reverse
  (bind-key "s-~" (lambda() () (interactive) (other-window -1)))

  ;; Because of the keybindings above, set one for `other-frame'
  (bind-key "S-1" 'other-frame)

  ;; Fullscreen!
  (setq ns-use-native-fullscreen nil) ; Not Lion style
  (bind-key "<s-return>" 'toggle-frame-fullscreen)

  ;; buffer switching
  (bind-key "M-[" 'previous-buffer)
  (bind-key "M-]" 'next-buffer)

  ;; Compiling
  (bind-key "H-c" 'compile)
  (bind-key "H-r" 'recompile)
  (bind-key "H-s" (defun save-and-recompile () (interactive) (save-buffer) (recompile)))

  ;; disable the key that minimizes emacs to the dock because I don't
  ;; minimize my windows
  ;; (global-unset-key (kbd "C-z"))

  ;; Not going to use these commands
  (put 'ns-print-buffer 'disabled t)
  (put 'suspend-frame 'disabled t))

(use-package exec-path-from-shell
  :if (memq window-system '(mac ns))
  :ensure t
  :init
  (exec-path-from-shell-initialize))

(use-package multiple-cursors
  :ensure t
  :defer t
  :bind (("C-S-c C-S-c" . mc/edit-lines)
         ("C->"         . mc/mark-next-like-this)
         ("C-<"         . mc/mark-previous-like-this)
         ("C-c C-<"     . mc/mark-all-like-this)
         ("C-!"         . mc/mark-next-symbol-like-this)
         ("s-d"         . mc/mark-all-dwim)))

(use-package projectile
  :ensure t
  :diminish projectile-mode
  :commands (projectile-mode projectile-switch-project)
  :init
  ;; (setq projectile-completion-system 'vertico)
  (setq projectile-indexing-method 'alien)
  (setq projectile-enable-caching t)
  (global-set-key (kbd "C-c p p") 'projectile-switch-project)
  (setq projectile-switch-project-action #'projectile-find-dir)
  :config
  (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
  (projectile-mode +1))

(use-package rainbow-delimiters
  :ensure t
  :diminish
  :hook (prog-mode . rainbow-delimiters-mode))

(use-package rainbow-mode
  :ensure t
  :diminish
  :hook (prog-mode . rainbow-mode)
  :config
  (setq rainbow-x-colors nil))

(use-package smooth-scrolling
  :ensure t)

(use-package scratch
  :ensure t
  :commands scratch)

(use-package persistent-scratch
  :ensure t
  :init
  (persistent-scratch-setup-default))

(use-package smartparens
  :ensure t
  :defer t
  :diminish smartparens-mode
  :hook (prog-mode . smartparens-mode))

(use-package super-save
  :ensure t
  :defer t
  :config
  ;; (super-save-mode +1)
  )

(use-package undo-tree
  :ensure t
  :defer t
  :diminish
  :init
  (global-undo-tree-mode))

(use-package which-key
  :ensure t
  :defer t
  :diminish which-key-mode
  :hook (after-init . which-key-mode))

(use-package savehist
  :ensure t
  :init
  (setq history-length 25)
  (savehist-mode))

(use-package vertico
  :ensure t
  :bind (("C-x M-r" . vertico-repeat)
         :map vertico-map
         ;; M- keys for changing view:
         ("M-v"     . vertico-multiform-vertical)
         ("M-g"     . vertico-multiform-grid)
         ("M-f"     . vertico-multiform-flat)
         ("M-r"     . vertico-multiform-reverse)
         ("M-u"     . vertico-multiform-unobtrusive))
  :init (vertico-mode 1)
  :config (progn
            (add-hook 'minibuffer-setup-hook #'vertico-repeat-save)
            (vertico-mouse-mode 1)
            (vertico-multiform-mode 1)
            (setq vertico-count 20) ;; number of candidates and also size of minibuffer
            (setq vertico-multiform-categories '((consult-grep buffer))
                  vertico-multiform-commands '((tmm-menubar flat)
                                               (tmm-shortcut flat)))))

(use-package vertico-directory
  :after vertico
  :ensure nil
  ;; More convenient directory navigation commands
  :bind (:map vertico-map
              ;; left/right arrows for changing directory:
              ("<right>"   . vertico-directory-enter)
              ("<left>"    . vertico-directory-delete-word)
              ("M-<left>"  . vertico-directory-delete-char))
  ;; Tidy shadowed file names
  :hook (rfn-eshadow-update-overlay . vertico-directory-tidy))



;; https://github.com/oantolin/orderless
(use-package orderless
  :ensure t
  :after vertico
  :config (progn
            (setq orderless-matching-styles '(orderless-regexp
                                              orderless-initialism
                                              orderless-prefixes)
                  orderless-component-separator #'orderless-escapable-split-on-space)

            ;; Use the built-in "partial-completion" style to complete
            ;; file inputs such as "/e/ni/co.nix" into
            ;; "/etc/nixos/configuration.nix".  The "basic" style is
            ;; needed to support the hostname completion in the TRAMP
            ;; inputs such as "/sshx:HOSTNAME".
            (setq completion-category-defaults nil
                  completion-category-overrides '((file (styles basic partial-completion))))
            (setq completion-styles '(orderless basic))))


;; https://github.com/minad/marginalia
(use-package marginalia
  :ensure t
  :after vertico
  ;; :demand t
  :config (marginalia-mode 1))
